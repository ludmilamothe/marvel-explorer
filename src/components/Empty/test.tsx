import { screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import Empty from '.';

const datas = {
  imagePath: '/images/no-search.svg',
  imageAlt:
    'A magnifying glass with sad face and with some bubbles in the background',
  children: 'Some text',
};

describe('<Empty />', () => {
  it('should render the empty correct', () => {
    renderWithProviders(<Empty {...datas} />);

    const image = screen.getByRole('img', { name: datas.imageAlt });
    const children = screen.getByText(datas.children);

    expect(children).toBeInTheDocument();

    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute('src', datas.imagePath);
  });

  it('should render the empty styles correct', () => {
    const { container } = renderWithProviders(<Empty {...datas} />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
