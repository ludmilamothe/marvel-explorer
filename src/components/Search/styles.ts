import styled, { css } from 'styled-components';
import media from 'styled-media-query';

export const Container = styled.form`
  ${({ theme }) => css`
    width: 100%;
    max-width: 40rem;
    position: relative;
    margin-top: ${theme.spacings.medium};
    margin-bottom: ${theme.spacings.large};

    ${media.greaterThan('medium')`
      max-width: 57rem;
      margin-top: calc(${theme.spacings.large} * 2);
      margin-bottom: ${theme.spacings.xlarge};
    `}
  `}
`;

export const Label = styled.label`
  ${({ theme }) => css`
    width: 100%;
    display: flex;
    flex-direction: column;
    font-size: ${theme.font.sizes.medium};
    color: ${theme.colors.white};

    ${media.greaterThan('medium')`
      font-size: ${theme.font.sizes.large};
    `}
  `}
`;

export const InputWrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    width: 100%;
    margin-top: ${theme.spacings.xxsmall};

    ${media.greaterThan('medium')`
      margin-top: ${theme.spacings.xsmall};
    `}
  `}
`;

export const Input = styled.input`
  ${({ theme }) => css`
    flex: 1;
    background-color: ${theme.colors.white};
    padding: ${theme.spacings.xsmall};
    color: ${theme.colors.black};
    font-size: ${theme.font.sizes.medium};
    border: 0;
    border-radius: ${theme.border.radius} 0 0 ${theme.border.radius};

    &::placeholder {
      color: ${theme.colors.black};
    }

    ${media.greaterThan('medium')`
      padding: ${theme.spacings.small};
      font-size: ${theme.font.sizes.large};
    `}
  `}
`;

export const Button = styled.button`
  ${({ theme }) => css`
    background-color: ${theme.colors.primary};
    padding: 0 ${theme.spacings.small};
    color: ${theme.colors.white};
    font-size: ${theme.font.sizes.medium};
    flex-shrink: 0;
    border: 0;
    border-radius: 0 ${theme.border.radius} ${theme.border.radius} 0;
    transition: opacity 0.2s;

    &:hover {
      opacity: 0.8;
    }

    ${media.greaterThan('small')`
      padding: 0 ${theme.spacings.medium};
    `}

    ${media.greaterThan('medium')`
      padding: 0 ${theme.spacings.large};
      font-size: ${theme.font.sizes.large};
    `}
  `}
`;

export const Error = styled.span`
  ${({ theme }) => css`
    position: absolute;
    top: calc(100% + ${theme.spacings.xxsmall});
    color: ${theme.colors.primary};
    font-size: ${theme.font.sizes.medium};

    ${media.greaterThan('medium')`
      font-size: ${theme.font.sizes.large};
      top: calc(100% + ${theme.spacings.xsmall});
    `}
  `}
`;
