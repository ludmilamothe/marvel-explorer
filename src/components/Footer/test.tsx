import { screen } from '@testing-library/react';
import { renderWithTheme } from 'utils/test/helpers';

import Footer from '.';

describe('<Footer />', () => {
  it('should render the footer content correct', () => {
    renderWithTheme(<Footer />);

    const linkText = 'Mauro de Souza';
    const copy = 'Data provided by Marvel. Â© 2020 MARVEL';

    const linkElement = screen.getByRole('link', { name: linkText });

    expect(linkElement).toHaveAttribute(
      'href',
      'https://www.linkedin.com/in/maurodesouzaa/'
    );

    expect(linkElement).toHaveAttribute('target', '_blank');

    expect(screen.getByText(copy)).toBeInTheDocument();
  });

  it('should render the footer styles correct', () => {
    const { container } = renderWithTheme(<Footer />);

    const footer = container.firstChild;

    expect(footer).toMatchSnapshot();
  });
});
