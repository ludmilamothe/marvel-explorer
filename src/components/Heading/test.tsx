import { screen } from '@testing-library/react';
import { renderWithTheme } from 'utils/test/helpers';

import theme from 'styles/theme/default';
import Heading from '.';

describe('<Heading />', () => {
  it('should render the heading with children', () => {
    const text = 'Just a text to test';

    renderWithTheme(<Heading>{text}</Heading>);

    const heading = screen.getByRole('heading', { name: text });

    expect(heading).toBeInTheDocument();

    expect(heading).toMatchInlineSnapshot(`
      .c0 {
        color: #FFF;
        font-weight: 900;
        font-size: 2.4rem;
        max-width: 55rem;
        text-align: center;
      }

      @media (min-width:768px) {
        .c0 {
          font-size: 4.0rem;
        }
      }

      <h1
        class="c0"
      >
        Just a text to test
      </h1>
    `);
  });

  it('should render the heading with correct styles', () => {
    const text = 'Just a text to test';

    renderWithTheme(<Heading>{text}</Heading>);

    const heading = screen.getByRole('heading', { name: text });

    expect(heading).toHaveStyle({
      fontSize: theme.font.sizes.large,
      fontWeight: theme.font.xbold,
      color: theme.colors.white,
    });

    expect(heading).toHaveStyleRule('font-size', theme.font.sizes.xlarge, {
      media: '(min-width: 768px)',
    });
  });
});
