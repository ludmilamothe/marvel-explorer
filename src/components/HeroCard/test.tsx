/* eslint-disable @typescript-eslint/no-var-requires */

import { fireEvent, screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import { storageKey } from 'hooks/useFavoriteHeroes';
import HeroCard from '.';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const useRouter = jest.spyOn(require('next/router'), 'useRouter');
const push = jest.fn();

useRouter.mockImplementation(() => ({
  push,
}));

const heroData = {
  heroId: 123,
  heroName: 'Iron man',
  heroThumbnail:
    'http://i.annihil.us/u/prod/marvel/i/mg/9/c0/527bb7b37ff55.jpg',
};

describe('<HeroCard />', () => {
  beforeEach(() => {
    window.localStorage.clear();
  });

  it('should render the hero card correct', () => {
    renderWithProviders(<HeroCard {...heroData} />);

    expect(
      screen.getByRole('img', { name: heroData.heroName })
    ).toBeInTheDocument();

    expect(
      screen.getByRole('heading', { name: heroData.heroName })
    ).toBeInTheDocument();

    expect(screen.getByLabelText(/add hero to favorites/i)).toBeInTheDocument();
  });

  it('should render the hero card styles correct', () => {
    const { container } = renderWithProviders(<HeroCard {...heroData} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should add hero to favorites', () => {
    renderWithProviders(<HeroCard {...heroData} />);

    const addToFavoriteElement = screen.getByLabelText(
      /add hero to favorites/i
    );

    expect(addToFavoriteElement).toBeInTheDocument();
    fireEvent.click(addToFavoriteElement);

    const removeFromFavoriteElement = screen.getByLabelText(
      /remove hero to favorites/i
    );

    expect(removeFromFavoriteElement).toBeInTheDocument();
    expect(addToFavoriteElement).not.toBeInTheDocument();

    const storagedHero = JSON.parse(window.localStorage.getItem(storageKey)!);

    expect(storagedHero).toEqual([heroData]);
  });

  it('should remove hero to favorites', () => {
    window.localStorage.setItem(storageKey, JSON.stringify([heroData]));

    renderWithProviders(<HeroCard {...heroData} />);

    const removeFromFavoriteElement = screen.getByLabelText(
      /remove hero to favorites/i
    );

    expect(removeFromFavoriteElement).toBeInTheDocument();
    fireEvent.click(removeFromFavoriteElement);

    const addToFavoriteElement = screen.getByLabelText(
      /add hero to favorites/i
    );

    expect(removeFromFavoriteElement).not.toBeInTheDocument();
    expect(addToFavoriteElement).toBeInTheDocument();

    const storagedHero = JSON.parse(window.localStorage.getItem(storageKey)!);
    expect(storagedHero).toEqual([]);
  });

  it('should navigate to hero page', () => {
    const { container } = renderWithProviders(<HeroCard {...heroData} />);

    const heroCard = container.firstChild!;
    fireEvent.click(heroCard);

    expect(push).toBeCalledTimes(1);
    expect(push).toBeCalledWith(`/characters/${heroData.heroId}`);
  });
});
