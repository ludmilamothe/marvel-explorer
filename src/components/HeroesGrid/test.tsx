import { renderWithProviders } from 'utils/test/helpers';

import HeroesGrid from '.';

describe('<HeroesGrid />', () => {
  it('should render the HeroesGrid styles correct', () => {
    const { container } = renderWithProviders(<HeroesGrid />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
