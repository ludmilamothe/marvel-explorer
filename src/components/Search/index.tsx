import { FormEvent, useCallback, useRef } from 'react';
import * as S from './styles';

export type SearchProps = {
  label?: string;
  placeholder?: string;
  error?: string;
  onSubmit?: (inputValue: string) => void;
};

const Search = ({
  label = 'Type the name of your hero',
  placeholder = 'Ex: iron main',
  error = '',
  onSubmit,
}: SearchProps) => {
  const inputRef = useRef<HTMLInputElement>(null);

  const handleSubmit = useCallback(
    (e: FormEvent) => {
      e.preventDefault();

      const inputValue = inputRef.current?.value || '';

      !!onSubmit && onSubmit(inputValue);
    },
    [onSubmit]
  );

  return (
    <S.Container onSubmit={handleSubmit}>
      <S.Label htmlFor="search">
        {label}

        <S.InputWrapper>
          <S.Input id="search" ref={inputRef} placeholder={placeholder} />
          <S.Button type="submit">Search</S.Button>
        </S.InputWrapper>
      </S.Label>

      {!!error && <S.Error>{error}</S.Error>}
    </S.Container>
  );
};

export default Search;
