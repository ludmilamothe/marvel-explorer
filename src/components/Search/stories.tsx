import { Meta, Story } from '@storybook/react/types-6-0';

import Search, { SearchProps } from '.';

export default {
  title: 'Search',
  component: Search,
} as Meta<SearchProps>;

export const Basic: Story<SearchProps> = args => <Search {...args} />;
