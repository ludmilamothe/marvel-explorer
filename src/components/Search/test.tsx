import { fireEvent, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { renderWithProviders } from 'utils/test/helpers';

import Search, { SearchProps } from '.';

const searchProps: Required<Omit<SearchProps, 'onSubmit' | 'error'>> = {
  label: 'label',
  placeholder: 'placeholder',
};

describe('<Search />', () => {
  it('should render the search correct', () => {
    renderWithProviders(<Search {...searchProps} />);

    const input = screen.getByPlaceholderText(searchProps.placeholder);
    const label = screen.getByLabelText(searchProps.label);
    const button = screen.getByRole('button', { name: /search/i });

    expect(input).toBeInTheDocument();
    expect(button).toBeInTheDocument();
    expect(label).toBeInTheDocument();
  });

  it('should render the search styles correct', () => {
    const { container } = renderWithProviders(<Search />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render error', () => {
    const errorText = 'some error';

    renderWithProviders(<Search error={errorText} />);

    expect(screen.getByText(errorText)).toBeInTheDocument();
  });

  it('should call onSubmit function when passed', () => {
    const onSubmit = jest.fn();
    const text = 'some value';

    renderWithProviders(<Search {...searchProps} onSubmit={onSubmit} />);

    const input = screen.getByPlaceholderText(searchProps.placeholder);
    const button = screen.getByRole('button', { name: /search/i });

    userEvent.type(input, text);
    fireEvent.click(button);

    expect(onSubmit).toBeCalledWith(text);
    expect(onSubmit).toBeCalledTimes(1);
  });

  it('should call onSubmit function with no value', () => {
    const onSubmit = jest.fn();

    renderWithProviders(<Search {...searchProps} onSubmit={onSubmit} />);

    const button = screen.getByRole('button', { name: /search/i });
    fireEvent.click(button);

    expect(onSubmit).toBeCalledWith('');
    expect(onSubmit).toBeCalledTimes(1);
  });
});
