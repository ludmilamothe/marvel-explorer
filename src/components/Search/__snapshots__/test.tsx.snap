// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> should render the search styles correct 1`] = `
.c0 {
  width: 100%;
  max-width: 40rem;
  position: relative;
  margin-top: 3.2rem;
  margin-bottom: 4.0rem;
}

.c1 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-size: 1.6rem;
  color: #FFF;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  margin-top: 0.8rem;
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  background-color: #FFF;
  padding: 1.2rem;
  color: #202020;
  font-size: 1.6rem;
  border: 0;
  border-radius: 0.4rem 0 0 0.4rem;
}

.c3::-webkit-input-placeholder {
  color: #202020;
}

.c3::-moz-placeholder {
  color: #202020;
}

.c3:-ms-input-placeholder {
  color: #202020;
}

.c3::placeholder {
  color: #202020;
}

.c4 {
  background-color: #E62429;
  padding: 0 2.4rem;
  color: #FFF;
  font-size: 1.6rem;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  border: 0;
  border-radius: 0 0.4rem 0.4rem 0;
  -webkit-transition: opacity 0.2s;
  transition: opacity 0.2s;
}

.c4:hover {
  opacity: 0.8;
}

@media (min-width:768px) {
  .c0 {
    max-width: 57rem;
    margin-top: calc(4.0rem * 2);
    margin-bottom: 4.8rem;
  }
}

@media (min-width:768px) {
  .c1 {
    font-size: 2.4rem;
  }
}

@media (min-width:768px) {
  .c2 {
    margin-top: 1.2rem;
  }
}

@media (min-width:768px) {
  .c3 {
    padding: 2.4rem;
    font-size: 2.4rem;
  }
}

@media (min-width:450px) {
  .c4 {
    padding: 0 3.2rem;
  }
}

@media (min-width:768px) {
  .c4 {
    padding: 0 4.0rem;
    font-size: 2.4rem;
  }
}

<form
  class="c0"
>
  <label
    class="c1"
    for="search"
  >
    Type the name of your hero
    <div
      class="c2"
    >
      <input
        class="c3"
        id="search"
        placeholder="Ex: iron main"
      />
      <button
        class="c4"
        type="submit"
      >
        Search
      </button>
    </div>
  </label>
</form>
`;
